### ISSUE TITLE
No module named 'mmcv._ext'

### ISSUE THREAD

**Issue by AshishGupta1992** [open-mmlab/mmdetection#3765]
![Issue](https://user-images.githubusercontent.com/5601603/93168558-d31bbf00-f740-11ea-9324-572e103dffea.PNG)
Thanks for your error report and we appreciate it a lot.

**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**

**Reproduction**
1. What command or script did you run?
```
A placeholder for the command.
```
2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?

**Environment**

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment information and paste it here.
2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error trackback here.
```
A placeholder for trackback.
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**Comment by yuzhj** [comment]
you can have a try
pip uninstall mmcv
pip install mmcv-full

**Comment by AshishGupta1992** [comment]
> you can have a try
> pip uninstall mmcv
> pip install mmcv-full

I already tried. It is only working on 1.0.5 version.When i install mmcv-full,it updates with the latest version which is 1.1.2 but the version i need to install is 1.0.5 .

And when i install mmcv-full=1.0.5, I get 

**ERROR: Could not find a version that satisfies the requirement mmcv-full==1.0.5 (from versions: none)
ERROR: No matching distribution found for mmcv-full==1.0.5**

**Comment by yuzhj** [comment]
> > you can have a try
> > pip uninstall mmcv
> > pip install mmcv-full
> 
> I already tried. It is only working on 1.0.5 version.When i install mmcv-full,it updates with the latest version which is 1.1.2 but the version i need to install is 1.0.5 .
> 
> And when i install mmcv-full=1.0.5, I get
> 
> **ERROR: Could not find a version that satisfies the requirement mmcv-full==1.0.5 (from versions: none) ERROR: No matching distribution found for mmcv-full==1.0.5**

Maybe you should git clone the mmcv repository and compile it by yourself

**Comment by bharath-07** [comment]
I am facing the same error. I am using v2.4 mmdetection and installed mmcv-full.
Any help is appreciated.

**Comment by ZwwWayne** [comment]
> > you can have a try
> > pip uninstall mmcv
> > pip install mmcv-full
> 
> I already tried. It is only working on 1.0.5 version.When i install mmcv-full,it updates with the latest version which is 1.1.2 but the version i need to install is 1.0.5 .
> 
> And when i install mmcv-full=1.0.5, I get
> 
> **ERROR: Could not find a version that satisfies the requirement mmcv-full==1.0.5 (from versions: none) ERROR: No matching distribution found for mmcv-full==1.0.5**

pip install mmcv-full==1.0.5 should work, I just tried. If you failed, please report your running environment, including OS, CUDA, pytorch version etc.

**Comment by ZwwWayne** [comment]
Please refer to the doc for [trouble shooting](https://github.com/open-mmlab/mmdetection/blob/master/docs/trouble_shooting.md).

**Comment by TheFebrin** [comment]
This issue pointed me towards the solution: https://github.com/open-mmlab/mmdetection/issues/3271

**Comment by ahmadkaleem2** [comment]
this happened to me on other versions of mmcv-full but this one fixed it.
try this 
#									                                                   {cuda_ver}/{torch_version}
pip install mmcv-full==1.4.5 -f https://download.openmmlab.com/mmcv/dist/cu113/torch1.10/index.html

**Comment by ahmadkaleem2** [comment]
afterwards
in terminal try:
python
from mmcv.ops import box_iou_rotated

**Comment by ejduwell** [comment]
For anyone still having trouble with the "no module named 'mmcv.ext' above:
I had this issue as well working in OSX on a new m1 macbook. I was able to solve the issue by building mmcv from source. There are detailed directions on how to do so here:

https://mmcv.readthedocs.io/en/latest/get_started/build.html

However, I've also pasted the steps below:

git clone https://github.com/open-mmlab/mmcv.git
cd mmcv

pip install -r requirements/optional.txt

!****"If you are on macOS, add the following environment variables before the installing command."****!:
CC=clang CXX=clang++ CFLAGS='-stdlib=libc++' MMCV_WITH_OPS=1 pip install -e .

I complied/installed mmcv in my /opt/anaconda3/lib/python3.9/site-packages/ directory along with mmocr. That way they are both automatically on my path in my anaconda/python environments.

Hope this helps,
Ethan

**Comment by pasqualecelardo** [comment]
Hi @ejduwell ,
I tried to perform the same procedure on my macbook m1, but I keep having the same problem. 
Do you know of any other solution?
Thank you in advance.

Regards
Pasquale

**Comment by ejduwell** [comment]
Hi @pasqualecelardo ,

Its been a while since I dealt with this one... I finished the project that originally required me to wrestle with this a while back. However, I just looked at my final install script and think I may know what’s going on. (or at the very least may have something for you to try..) I think mmocr/openmmlabs may have moved  to using (or assuming developers will be using) their own package installer called “openmim” for installing packages like mmcv and mmdet, etc.. I think I moved to using that instead.

Here’s a link to the repository I developed that used mmocr: https://github.com/ejduwell/TextyBeast#readme

If you want to see the way I ended up doing the mmocr environment install on Macs check out the ‘buildEnvs_Mac.sh’ script. The section you’re probably interested in is from lines 77-108:
https://github.com/ejduwell/TextyBeast/blob/a33ff030b38f3b9fe9e359715ed01bef40d34d88/buildEnvs_Mac.sh#L77

In short, it looks like I ended up installing using the following:

	pip install -U openmim
	mim install mmengine
	mim install mmcv==1.7.1
	mim install mmdet==2.28.2


Hope this helps!

Best,
Ethan

**Comment by pasqualecelardo** [comment]
hi @ejduwell , 
thank you for your response. 
I will take a look at your script now. 
I will add a note, on windows and ubuntu I have not had these problems. 
In fact I installed mmcv 2.0.1 without the help of openmim. 
Regards
Pasquale

**Comment by ejduwell** [comment]
Hi @pasqualecelardo ,
No problem. Hope that helps. I recall having mac-specific issues too. I ended up needing to have two different scripts for setting up the environment, one for ubuntu and another for macos. (hence the mac-specific ‘buildEnvs_Mac.sh’ install script above). My strategy was to basically detect the operating system in the install process, and then use separate environment setup scripts for each. Perhaps there is a better way, but that worked for my purposes.

-Ethan
